blob/s3blob: nil pointer dereference for range reader error path

### Describe the bug

Panic caused by nil pointer dereference in blob Reader after error in `Read`.
Debugging this it seems this is caused by `Reader` entering an invalid state:
1. `Read` is called and we reach line 147 (in version `0.37.0`) and `NewRangeReader` returns an error:
```
			r.r, err = r.b.NewRangeReader(r.ctx, r.key, r.baseOffset+r.relativeOffset, length, r.dopts)
			if err != nil {
				return 0, wrapError(r.b, err, r.key)
			}
```
2. r.r is now `nil`.
3. `Seek` (**which doesn't validate the object**) is later called for the reader, we reach `Size` and `r.r.Attributes().Size` panics.

### To Reproduce

Managed to reproduce this 100% of the time in a nonsensical way with context canceling:
```
package main

import (
	"context"
	"io"

	awsv2cfg "github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	s3v2 "github.com/aws/aws-sdk-go-v2/service/s3"
	"gocloud.dev/blob"
	"gocloud.dev/blob/s3blob"
)

const (
	AWS_ACCESS_KEY_ID     = "XXXXX"
	AWS_SECRET_ACCESS_KEY = "XXXXX"
	AWS_SESSION_TOKEN     = "XXXXX"
)

func main() {
	ctx, cancelFunc := context.WithCancel(context.Background())

	client := openClient(ctx)
	bucket, err := s3blob.OpenBucketV2(ctx, client, "test", nil)
	if err != nil {
		panic(err)
	}

	reader, err := bucket.NewRangeReader(ctx, "test.yaml", 0, 1024, &blob.ReaderOptions{})
	if err != nil {
		panic(err)
	}

	b := make([]byte, 512)
	_, err = reader.Read(b)
	if err != nil {
		panic(err)
	}

	_, err = reader.Seek(0, io.SeekStart)
	if err != nil {
		panic(err)
	}

	cancelFunc()

	_, _ = reader.Read(b)

	_, err = reader.Seek(0, io.SeekEnd)
	if err != nil {
		panic(err)
	}
}

func openClient(ctx context.Context) *s3v2.Client {
	funcs := setAWSLoadOptions()
	cfg, err := awsv2cfg.LoadDefaultConfig(ctx, funcs...)
	if err != nil {
		panic(err)
	}

	return s3v2.NewFromConfig(cfg)
}

func setAWSLoadOptions() []func(*awsv2cfg.LoadOptions) error {
	funcs := []func(*awsv2cfg.LoadOptions) error{awsv2cfg.WithRegion("us-east-1")}
	funcs = append(funcs, awsv2cfg.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(
		AWS_ACCESS_KEY_ID,
		AWS_SECRET_ACCESS_KEY,
		AWS_SESSION_TOKEN)))

	return funcs
}
```

Traceback:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x20 pc=0x100e5f840]

goroutine 1 [running]:
gocloud.dev/blob.(*Reader).Size(0x140004c0000)
        /Users/user/go/pkg/mod/gocloud.dev@v0.37.0/blob/blob.go:222 +0x140
gocloud.dev/blob.(*Reader).Seek(0x140004c0000, 0x0, 0x2)
        /Users/user/go/pkg/mod/gocloud.dev@v0.37.0/blob/blob.go:169 +0x40
main.main()
        /Users/user/dev/example.go:49 +0x128

Process finished with the exit code 2
```

## Expected behavior

Expected an error and not to panic.

### Version

gocloud 0.37.0

### Additional context

The original program consists of a lot of goroutines working on the reader with locks, the use itself is not ideal but at the very least it probably shouldn't crash. I would expect that if one gorountines errors (say from timeout getting context deadline exceeded) another goroutine using it would also get an error and not panic.

--- a/blob/blob.go
+++ b/blob/blob.go
@@ -128,12 +128,11 @@ func (r *Reader) Read(p []byte) (int, error) {
 		// to (SeekEnd, 0) and use the return value to determine the size
 		// of the data, then Seek back to (SeekStart, 0).
 		saved := r.savedOffset
-		r.savedOffset = -1
 		if r.relativeOffset == saved {
 			// Nope! We're at the same place we left off.
+			r.savedOffset = -1
 		} else {
 			// Yep! We've changed the offset. Recreate the reader.
-			_ = r.r.Close()
 			length := r.baseLength
 			if length >= 0 {
 				length -= r.relativeOffset
@@ -142,11 +141,13 @@ func (r *Reader) Read(p []byte) (int, error) {
 					return 0, gcerr.Newf(gcerr.Internal, nil, "blob: invalid Seek (base length %d, relative offset %d)", r.baseLength, r.relativeOffset)
 				}
 			}
-			var err error
-			r.r, err = r.b.NewRangeReader(r.ctx, r.key, r.baseOffset+r.relativeOffset, length, r.dopts)
+			newR, err := r.b.NewRangeReader(r.ctx, r.key, r.baseOffset+r.relativeOffset, length, r.dopts)
 			if err != nil {
 				return 0, wrapError(r.b, err, r.key)
 			}
+			_ = r.r.Close()
+			r.savedOffset = -1
+			r.r = newR
 		}
 	}
 	n, err := r.r.Read(p)

--- a/blob/blob_test.go
+++ b/blob/blob_test.go
@@ -265,6 +265,74 @@ func TestDownloader(t *testing.T) {
 	}
 }
 
+func TestSeekAfterReadFailure(t *testing.T) {
+	const filename = "f.txt"
+
+	ctx := context.Background()
+
+	bucket := NewBucket(&oneTimeReadBucket{first: true})
+	defer bucket.Close()
+
+	reader, err := bucket.NewRangeReader(ctx, filename, 0, 100, nil)
+	if err != nil {
+		t.Fatalf("failed NewRangeReader: %v", err)
+	}
+	defer reader.Close()
+
+	b := make([]byte, 10)
+
+	_, err = reader.Read(b)
+	if err != nil {
+		t.Fatalf("failed Read#1: %v", err)
+	}
+
+	_, err = reader.Seek(0, io.SeekStart)
+	if err != nil {
+		t.Fatalf("failed Seek#1: %v", err)
+	}
+
+	// This Read will force a recreation of the reader via NewRangeReader,
+	// which will fail.
+	_, err = reader.Read(b)
+	if err == nil {
+		t.Fatalf("unexpectedly succeeded Read#2: %v", err)
+	}
+
+	_, err = reader.Seek(0, io.SeekStart)
+	if err != nil {
+		t.Fatalf("failed Seek#2: %v", err)
+	}
+}
+
+// oneTimeReadBucket implements driver.Bucket for TestSeekAfterReadFailure.
+// It returns a fake reader that succeeds once, then fails.
+type oneTimeReadBucket struct {
+	driver.Bucket
+	first bool
+}
+
+type workingReader struct {
+	driver.Reader
+}
+
+func (r *workingReader) Read(p []byte) (int, error) {
+	return len(p), nil
+}
+
+func (r *workingReader) Attributes() *driver.ReaderAttributes { return &driver.ReaderAttributes{} }
+func (r *workingReader) Close() error                         { return nil }
+
+func (b *oneTimeReadBucket) NewRangeReader(ctx context.Context, key string, offset, length int64, opts *driver.ReaderOptions) (driver.Reader, error) {
+	if b.first {
+		b.first = false
+		return &workingReader{}, nil
+	}
+	return nil, errFake
+}
+
+func (b *oneTimeReadBucket) ErrorCode(err error) gcerrors.ErrorCode { return gcerrors.Unknown }
+func (b *oneTimeReadBucket) Close() error                           { return nil }
+
 // erroringBucket implements driver.Bucket. All interface methods that return
 // errors are implemented, and return errFake.
 // In addition, when passed the key "work", NewRangeReader and NewTypedWriter

